music21/musicxml/m21ToXml.py

```diff
@@ -4341,7 +4415,9 @@ def noteToNotations(self, n, noteIndexInChord=0, chordParent=None):
                     applicableArticulations.append(a)
                 fingeringNumber += 1
             elif isSingleNoteOrFirstInChord:
-                applicableArticulations.append(a)
+                # hammer-on/pull-off are written from their Spanner representation
+                if not isinstance(a, (articulations.HammerOn, articulations.PullOff)):
+                    applicableArticulations.append(a)
```

Comment:
Should add a blaring comment something like "safe to ignore any HammerOn or PullOff a user might have put in this array"